src/algebra/smul_with_zero.lean

```diff
@@ -26,31 +26,59 @@ We also add an `instance`:
 * any `monoid_with_zero` has a `mul_action_with_zero R R` acting on itself.
 -/
 
-variables (R M : Type*)
+variables {R M M' : Type*}
 
 section has_zero
 
-variable [has_zero M]
+variables [has_zero R] [has_zero M]
 
+variables (R M)
 /--  `smul_with_zero` is a class consisting of a Type `R` with `0 ∈ R` and a scalar multiplication
 of `R` on a Type `M` with `0`, such that the equality `r • m = 0` holds if at least one among `r`
 or `m` equals `0`. -/
-class smul_with_zero [has_zero R] extends has_scalar R M :=
+class smul_with_zero extends has_scalar R M :=
 (smul_zero : ∀ r : R, r • (0 : M) = 0)
 (zero_smul : ∀ m : M, (0 : R) • m = 0)
 
-variables {M}
-
-@[simp] lemma zero_smul [has_zero R] [smul_with_zero R M] (m : M) :
+variables (R) {M}
```

Comment:
```suggestion
variables {M}
```
I guess that `R` is already explicit, right?

Reply:
This was deliberate, I wanted to make it super clear what the explicitness was rather than having the reader have to piece things back together.